File: ruby/circular-buffer/circular_buffer.rb
class CircularBuffer
  BufferEmptyException = Class.new(StandardError)
  BufferFullException  = Class.new(StandardError)

  attr_reader :buffer, :size

  def initialize(size)
    clear(size)
  end

  def write(item)
    return if item.nil?
    raise BufferFullException unless buffer[@write].nil?
    write_buffer(item)
  end

  def write!(item)
    return if item.nil?
    next_read unless buffer[@write].nil?
    write_buffer(item)
  end

  def write_buffer(item)
    buffer[@write] = item
    next_write
  end

  def read
    raise BufferEmptyException if buffer.compact.empty?
    item = buffer[@read]

    buffer[@read] = nil
    next_read

    item
  end

  def clear(size = buffer.size)
    @buffer = Array.new(size)
    @size   = size
    @write  = 0
    @read   = 0
  end

  private

  def next_read
    @read = ((@read + 1) % size)
  end

  def next_write
    @write = ((@write + 1) % size)
  end
end

File: go/circular-buffer/circular_buffer.go
package circular

import "errors"

const testVersion = 3

type Buffer struct {
	read   int
	write  int
	array  []byte
	length int
}

func NewBuffer(size int) *Buffer {
	return &Buffer{array: make([]byte, size)}
}

func (buffer *Buffer) ReadByte() (byte, error) {
	if buffer.length == 0 {
		return 0, errors.New("Buffer is empty")
	}
	result := buffer.array[buffer.read]
	buffer.nextRead()
	buffer.length--
	return result, nil
}

func (buffer *Buffer) WriteByte(c byte) error {
	if buffer.full() {
		return errors.New("Buffer is full")
	}
	buffer.array[buffer.write] = c
	buffer.nextWrite()
	buffer.length++
	return nil
}

func (buffer *Buffer) Overwrite(c byte) {
	if buffer.full() {
		buffer.nextRead()
	} else {
		buffer.length++
	}
	buffer.array[buffer.write] = c
	buffer.nextWrite()
}

func (buffer *Buffer) Reset() {
	buffer.read = 0
	buffer.write = 0
	buffer.length = 0
}

func (buffer *Buffer) full() bool {
	return buffer.length >= len(buffer.array)
}

func (buffer *Buffer) nextRead() {
	buffer.read = (buffer.read + 1) % len(buffer.array)
}

func (buffer *Buffer) nextWrite() {
	buffer.write = (buffer.write + 1) % len(buffer.array)
}
